src/app/decoder.cpp

```diff
@@ -55,39 +40,26 @@ uint16_t extractApsFrame(uint8_t * buffer, uint16_t buf_length, EmberApsFrame *
         return 0;
     }
 
-    const uint8_t * read_ptr = buffer;
+    chip::DataModelReader reader(buffer, buf_length);
 
-    // Skip first byte, because that's the always-0 frame control.
-    ++read_ptr;
-    --buf_length;
-
-#define TRY_READ(fieldName, fieldSize)                                                                                             \
-    do                                                                                                                             \
-    {                                                                                                                              \
-        static_assert(sizeof(outApsFrame->fieldName) == fieldSize, "incorrect size for " #fieldName);                              \
-        if (buf_length < fieldSize)                                                                                                \
-        {                                                                                                                          \
-            ChipLogError(Zcl, "Missing " #fieldName " when extracting APS frame");                                                 \
-            return 0;                                                                                                              \
-        }                                                                                                                          \
-        outApsFrame->fieldName = Reader<fieldSize>::read(read_ptr);                                                                \
-        buf_length             = static_cast<uint16_t>(buf_length - fieldSize);                                                    \
-    } while (0)
-
-    TRY_READ(profileId, 2);
-    TRY_READ(clusterId, 2);
-    TRY_READ(sourceEndpoint, 1);
-    TRY_READ(destinationEndpoint, 1);
-    TRY_READ(options, 2);
-    TRY_READ(groupId, 2);
-    TRY_READ(sequence, 1);
-    TRY_READ(radius, 1);
+    CHIP_ERROR err = CHIP_NO_ERROR;
 
-#undef TRY_READ
-
-    // Cast is safe because buf_length is uint16_t, so we can't have moved too
-    // far along.
-    return static_cast<uint16_t>(read_ptr - buffer);
+    // Skip first byte, because that's the always-0 frame control.
+    uint8_t ignored;
+    err = reader.ReadOctet(&ignored)
```

Comment:
I love how readable this is! :heart:

Should we have a SkipBytes() call to avoid the ignored?

Reply:
The ignored is an artifact of our encoding right now being something we just made up that has nothing to do with anything, really.  In the long run we won't have a byte we're ignoring here.

But yes, if we end up with "real" code that needs to do that, we can add that.